### PR TITLE
PKCS12 support [work in progress]

### PR DESCRIPTION
**Pull request by haneric21**
https://github.com/cloudflare/cfssl/wiki/PKCS-%2312-Support-Requirements

Trying to parse the key and cert from gencert in pkcs12.go, then calling AGWA-forks/go-pkcs12's encode to output pkcs12.
...Still unable to correctly convert key []byte to interface{} and cert []byte to *x509.Certificate, which is required in Encode function.

### CODE REVIEW

**Comment by haneric21** [comment]
Successfully(I think) printing out PKCS12 file with command "cfssl gencert -initca -out-file pkcs12 csr.json" with no password.
Working on adding optional password parameter.

File: cli/cli.go [hunk 1]
```diff
@@ -173,6 +173,32 @@ func PrintCert(key, csrBytes, cert []byte) {
 	fmt.Printf("%s\n", jsonOut)
 }
 
+// PrintCert outputs a cert, key, csr, and outfile to stdout
```
Comment (kisom) [inline cli/cli.go]: Comment should reference `PrintCertAndFile`.

File: cli/gencert/gencert.go [hunk 2]
```diff
@@ -83,7 +85,19 @@ func gencertMain(args []string, c cli.Config) error {
 			}
 
 		}
-		cli.PrintCert(key, csrPEM, cert)
+		if c.OutFile == "pkcs12" {
+			password := make([]byte, 0)
```
Comment (kisom) [inline cli/gencert/gencert.go]: You don't really need to call `make` here. You can do `var password []byte`.

File: cli/gencert/gencert.go [hunk 3]
```diff
@@ -83,7 +85,19 @@ func gencertMain(args []string, c cli.Config) error {
...
+			if c.Password != "0" {
+				password = []byte(c.Password)
+			}
+			pkcs12Key, err := pkcs12.ParseAndEncode(key, cert, password)
+			if err != nil {
+				log.Errorf("%v\n", err)
+			}
+			cli.PrintCertAndFile(key, csrPEM, cert, pkcs12Key)
```
Comment (kisom) [inline cli/gencert/gencert.go]: An exported PKCS12 key is binary data, so you might not want to dump it to standard output.

File: crypto/pkcs12/pkcs12.go [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+func ParseAndEncode(key, cert, password []byte) ([]byte, error) {
+	block, _ := pem.Decode(key)
+	privKey := block.Bytes
+	privateKey, _ := x509.ParsePKCS1PrivateKey(privKey)
+
+	certBlock, _ := pem.Decode(cert)
+	certBytes := certBlock.Bytes
+	certificate, err := x509.ParseCertificate(certBytes)
+
+	none := make([]*x509.Certificate, 0)
```
Comment (jkroll-cf) [inline crypto/pkcs12/pkcs12.go]: Here, too, you can use `var none []*x509.Certificate` instead of `make`.

**Comment by grittygrease** [comment]
@haneric21 Three things:
1) Please address the comments.
2) Provide an update on how the password feature is coming
3) Fix the issues exposed via CI: https://travis-ci.org/cloudflare/cfssl/jobs/87802134, test.sh runs `go vet` and `golint` which can help expose the issues.

**Comment by haneric21** [comment]
Addressed comments and finished optional password.
Currently working on implementing pkcs12 using ECPrivateKey and in Serve, as well as fixing issues exposed by CI

**Comment by grittygrease** [comment]
These are the two errors:
crypto/pkcs12/pkcs12.go:11:1: exported function ParseAndEncode should have comment or be unexported
crypto/pkcs12/pkcs12.go:27:1: exported function Encode should have comment or be unexported

**Comment by haneric21** [comment]
addressed commenting issues, added usage text for gencert, added error handling for pkcs12.go, and implemented pkcs12 output feature in generator.go

**Comment by kisom** [comment]
The tests are failing because you aren't pulling `github.com/AGWA-forks/go-pkcs12` in; this should be vendored using godep (as per #308).

**Pull request closed without merging**